### PR TITLE
#10121: migrate indexed fill op into ttnn

### PR DESCRIPTION
**Pull request by bharane-ab**
### Ticket
Link to Github Issue #10121 

### Problem description
Migrate indexed_fill op into TTNN

### What's changed

Migrated indexed_fill op to ttnn/operations/data_movement

### Checklist
- [x] [Post commit CI passes](https://github.com/tenstorrent/tt-metal/actions/runs/10284810375)
- [x] [Model regression CI testing passes](https://github.com/tenstorrent/tt-metal/actions/runs/10284831346)
- [x] [post commit c++ test](https://github.com/tenstorrent/tt-metal/actions/runs/10284813564)
- [x] [model test ](https://github.com/tenstorrent/tt-metal/actions/runs/10284818791)
- [x] [device regression](https://github.com/tenstorrent/tt-metal/actions/runs/10284829124)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/data_movement/indexed_fill/indexed_fill_pybind.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+
+#include "indexed_fill_pybind.hpp"
+#include "indexed_fill.hpp"
+#include "ttnn/cpp/pybind11/decorators.hpp"
+
+
+namespace ttnn::operations::data_movement {
+namespace detail {
+namespace py = pybind11;
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/indexed_fill/indexed_fill_pybind.cpp:17]: Please follow the ccl/all_gather example:
* Remove heavy includes, use pybind_fwd instead
* Move implementation to cpp

File: ttnn/cpp/ttnn/operations/data_movement/indexed_fill/indexed_fill_pybind.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                py::kw_only(),
+                py::arg("memory_config") = std::nullopt,
+                py::arg("queue_id") = 0});
+}
+
+}  // detail
+
+void bind_indexed_fill(py::module& module) {
+   detail::bind_indexed_fill(module);
+}
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/indexed_fill/indexed_fill_pybind.cpp:68]: you can remove this :D

File: ttnn/cpp/ttnn/operations/data_movement/indexed_fill/indexed_fill_pybind.cpp (hunk 1 above)
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/data_movement/indexed_fill/indexed_fill_pybind.cpp:17]: You can remove this and just use pybind11 as it is used in a single place

**Review by eyonland — DISMISSED** [review]
Please have this reviewed by an MCW colleague prior to posting the PR.

File: ttnn/cpp/ttnn/deprecated/tt_lib/csrc/tt_lib_bindings_tensor_dm_ops.cpp [hunk 3]
```diff
@@ -392,11 +391,7 @@ namespace tt::tt_metal::detail{
             R"doc(Converts a tensor sharded one way to another way)doc"
         );
 
-        //MOE ops
-        m_tensor.def("indexed_fill", &indexed_fill,
-            py::arg("batch_id"), py::arg("input_a"), py::arg("input_b"), py::arg("output_mem_config").noconvert() = operation::DEFAULT_OUTPUT_MEMORY_CONFIG, py::arg("dim") = 0,
-            R"doc(Replaces batch of input in input_b denoted by batch_ids into input_a)doc"
-        );
+        // //MOE ops
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/deprecated/tt_lib/csrc/tt_lib_bindings_tensor_dm_ops.cpp:394]: remove

File: ttnn/cpp/ttnn/operations/data_movement/indexed_fill/indexed_fill_pybind.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                const std::optional<ttnn::MemoryConfig>& memory_config,
+                uint8_t queue_id) {
+                    return self(queue_id, batch_id, input_tensor_a, input_tensor_b, memory_config);
+                },
+                pybind11::arg("batch_id").noconvert(),
+                pybind11::arg("input_tensor_a").noconvert(),
+                pybind11::arg("input_tensor_b").noconvert(),
+                pybind11::kw_only(),
+                pybind11::arg("memory_config") = std::nullopt,
+                pybind11::arg("queue_id") = 0});
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/data_movement/indexed_fill/indexed_fill_pybind.cpp:60]: py::arg("dim") = 0 is missing. please compare with the  tt_lib binding

File: ttnn/cpp/ttnn/operations/data_movement/indexed_fill/indexed_fill.hpp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+namespace operations {
+namespace data_movement {
+
+struct IndexedFillOperation {
+    static ttnn::Tensor operator()(
+        uint8_t queue_id,
+        const ttnn::Tensor& batch_id,
+        const ttnn::Tensor& input_tensor_a,
+        const ttnn::Tensor& input_tensor_b,
+        const std::optional<ttnn::MemoryConfig>& memory_config = std::nullopt);
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/data_movement/indexed_fill/indexed_fill.hpp:19]: dim parameter is missing

File: ttnn/cpp/ttnn/operations/data_movement/indexed_fill/indexed_fill.cpp [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+#include "ttnn/operations/data_movement/indexed_fill/indexed_fill.hpp"
+#include "ttnn/operations/data_movement/indexed_fill/device/indexed_fill_op.hpp"
+#include "ttnn/multi_device.hpp"
+#include "ttnn/common/constants.hpp"
+
+namespace ttnn::operations::data_movement{
+
+ttnn::Tensor IndexedFillOperation::operator()(uint8_t queue_id, const ttnn::Tensor& batch_id, const ttnn::Tensor& input_tensor_a, const ttnn::Tensor& input_tensor_b, const std::optional<ttnn::MemoryConfig>& memory_config) {
+    std::int64_t dim=0;
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/data_movement/indexed_fill/indexed_fill.cpp:13]: this should be a part of input params with an assert.

**Review by KalaivaniMCW — APPROVED** [review]
Review comments have been addressed. Approved.

**Review by ayerofieiev-tt — APPROVED** [review]
nice and clean. thank you!

**Review by ntarafdar — APPROVED** [review]

**Pull request merged**